GoogleConnector extended of possibility creating team drives.

jancusp opened the pull request:
- parse CSV file to List of own class consisting of teamDrive and List of Users
- create, update, delete of team drive implemented.

Not sure about update, because it is about visual elements of team drive.
Probably it isn't needed.

Commit:
GoogleConnector extended of possibility creating team drives.

- parse CSV file to List of own class consisting of teamDrive and List of Users
- create, update, delete of team drive implemented.

Not sure about update, because it is about visual elements of team drive.
Probably it isn't needed.

Commit:
GoogleConnector extended of possibility creating team drives.

- parse CSV file to List of own class consisting of teamDrive and List of Users
- create, update, delete of team drive implemented.

Not sure about update, because it is about visual elements of team drive.
Probably it isn't needed.

zlamalp reviewed (changes requested):
Please convert all indents to tabs instead of spaces. We use tabs in all sources.

Commit:
GoogleConnector extended of possibility creating team drives.

- parse CSV file to List of own class consisting of teamDrive and List of Users
- create, update, delete of team drive implemented.

Not sure about update, because it is about visual elements of team drive.
Probably it isn't needed.

Commit:
GoogleConnector extended of possibility creating team drives.

- parse CSV file to List of own class consisting of teamDrive and List of Users
- create, update, delete of team drive implemented.

Not sure about update, because it is about visual elements of team drive.
Probably it isn't needed.

Indents converted with reformat code function.

Spaces replaced with tab

Spaces replaced with tab

Commit:
Indents converted with reformat code function.

Commit:
Indents corrected

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsConnectionImpl.java:
```diff
@@ -148,4 +149,92 @@ private static Credential authorize() {
+		return null;
+	}*/
+	@Override
+	public Drive getDriveService() throws IOException {
```
Please move it to near ``getDirectoryService()``.

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsConnectionImpl.java:
```diff
@@ -148,4 +149,92 @@ private static Credential authorize() {
+	public Drive getDriveService() throws IOException {
+		Credential credential = authorize();
+		return new Drive.Builder(
+				HTTP_TRANSPORT, JSON_FACTORY, credential)
```
You can directly call ``authorize()`` in ``new Drive.Builder()``.

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsServiceImpl.java:
```diff
@@ -800,4 +826,286 @@ private void deleteMember(String groupName, String memberId) throws GoogleGroups
+	/**
+	 * New class representing tuple of team drive and its users from CSV file.
+	 */
+	public class DriveWithMembers {
```
You don't need to create class, because this is a basically ``Map<TeamDrive,List<User>>``. Since this doesn't contains any additional methods or specific ``equals()`` etc., please use the Map. You can get keys, values or their pairs the same way like this list of ``DriveWithMembers``.

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsServiceImpl.java:
```diff
@@ -800,4 +826,286 @@ private void deleteMember(String groupName, String memberId) throws GoogleGroups
+
+					if (!Objects.equals(dm.getTeamDrive().getBackgroundImageFile(), td.getBackgroundImageFile()) ||
+							!Objects.equals(dm.getTeamDrive().getColorRgb(), td.getColorRgb()) ||
+							!Objects.equals(dm.getTeamDrive().getThemeId(), td.getThemeId())) {
```
Based on docs for ``Drive.create()`` themeId and backgroundFile/colorRgb params can't be set together, but this implementation allows it if changed in a source. But you don't have these properties in an input file, so there is no necessity to update them from Perun. Changing name is not possible from Perun side, since Perun generates new state and old name is lost -> new is created, old deleted. Hence we probably don't need update method.

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsServiceImpl.java:
```diff
@@ -800,4 +826,286 @@ private void deleteMember(String groupName, String memberId) throws GoogleGroups
+					TeamDrive teamDriveResult = new TeamDrive();
+					List<User> userListResult = new ArrayList<>();
+
+					teamDriveResult.setName(line[0].toLowerCase());
```
Why do you call ``lowerCase()`` on team drive name ? Is there a constraint on Google side ?

zlamalp commented on src/main/java/cz/metacentrum/perun/googlegroupconnector/GoogleGroupsServiceImpl.java:
```diff
@@ -800,4 +826,286 @@ private void deleteMember(String groupName, String memberId) throws GoogleGroups
+					for (String userMail : membersEmail) {
+						userMail = userMail.replaceAll("\\s+", "");
+						User user = new User();
+						user.setPrimaryEmail(userMail);
```
Perun can provide Google with either ``primaryEmail`` od ``ID`` of user (based on what is available). Can Drive API work with both when inserting permissions for user on Drive ? If yes, please use property based on settings, like when we insert members to the group. If it requires only primaryEmail, then we should discuss it / test it, because CESNET instance provides only IDs of google accounts and not email address.

jancusp replied:
As i was looking into documentation, permission has field only for email of specific user. It looks like it will be necessary to test it.

Commit:
getDriveService() moved near to getDirectoryService().
Authorize() in getDriveService() called directly.
Used Map instead of own class insisted of TeamDrive and List of Users.
lowerCase() removed from TeamDrive.setName() method.
Update of Team Drive removed, not needed anymore.

Commit:
Readme file updated.
parser for teamDrives moved near to the other parsers.
